### ISSUE TITLE
Unhandled expression type: VAR

### ISSUE THREAD

**Issue by Raghul-S-1994** [google/closure-compiler#2832]
_Object destruction_ declaration in _for_ loop creates a runtime exception.

```javascript
function random_function() {
    var content = {length: 3};
    for (let k = 0, { length } = content; k < length; k++) {  
    }
}
```

The resulting runtime exception is

```
java.lang.RuntimeException: INTERNAL COMPILER ERROR.
Please report this problem.

Unhandled expression type: VAR
In scope: random_function() with root: FUNCTION random_function 328 [length: 125] [source_file: test.js] [change_time: 854]
	at com.google.javascript.jscomp.NewTypeInference.analyzeExprFwd(NewTypeInference.java:1683)
	at com.google.javascript.jscomp.NewTypeInference.analyzeConditionalStmFwd(NewTypeInference.java:1057)
	at com.google.javascript.jscomp.NewTypeInference.analyzeFunctionFwd(NewTypeInference.java:989)
	at com.google.javascript.jscomp.NewTypeInference.analyzeFunction(NewTypeInference.java:834)
	at com.google.javascript.jscomp.NewTypeInference.process(NewTypeInference.java:557)
	at com.google.javascript.jscomp.PhaseOptimizer$NamedPass.process(PhaseOptimizer.java:303)
	at com.google.javascript.jscomp.PhaseOptimizer.process(PhaseOptimizer.java:230)
	at com.google.javascript.jscomp.Compiler.check(Compiler.java:1079)
	at com.google.javascript.jscomp.Compiler.performChecksAndTranspilation(Compiler.java:872)
	at com.google.javascript.jscomp.Compiler.access$000(Compiler.java:103)
	at com.google.javascript.jscomp.Compiler$2.call(Compiler.java:806)
	at com.google.javascript.jscomp.Compiler$2.call(Compiler.java:803)
	at com.google.javascript.jscomp.CompilerExecutor$2.call(CompilerExecutor.java:101)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: Unhandled expression type: VAR
	... 17 more
```

**Comment by anonyco** [comment]
After some experimenting with the online compiler, I believe I found the reason. However, I am no java coder, so someone else will have to implement these changes. When running with simple optimizations, it errors. When running with advanced optimizations, the online closure compiler instead gives the following warning message.

```
JSC_UNREACHABLE_CODE: unreachable code at line 3 character 22
    for (let k = 0, { length } = content; k < length; k++) {
                      ^
JSC_TYPE_MISMATCH: right side of numeric comparison
found   : undefined
required: number at line 3 character 42
    for (let k = 0, { length } = content; k < length; k++) {
                                          ^
JSC_UNREACHABLE_CODE: unreachable code at line 3 character 59
    for (let k = 0, { length } = content; k < length; k++) {
```

I believe the problem is that somewhere in the closure compiler coding, someone erroneously forgot to implement the recursive type checking for that kind of declaration. Basically, in layman's terms, when the compiler sees this: `for (let { foo } = {foo: "bar"};;);`, closure compiler does NOT thinking that foo is "bar", so it is a string. Rather, the closure compiler essentially ignored the curly brackets around foo and instead assumes foo is the right hand side of the expression (`{foo: "bar"}`), which is an object value. But, objects cannot be compared without type coercion. So, more evaluation must be done to optimize it. Then during more optimization somewhere else in the closure compiler, they probably coded it correctly and it finds foo to be a string. THEN, the place in the closure compiler code thinking that foo is a string  conflicts with another place that thinks that foo is an object, they clash heads, and the whole thing errors happily ever after.

P.S. If this is just a simple mistake, no one should give the closure compiler people a hard time about this. Honestly, it's pretty amazing how few bugs like this they introduced into a project of this size. All this bug really proves is that the closure compiler people are still humans who make mistakes just like the rest of us.

**Comment by brad4d** [comment]
Looks like the code above actually triggers a transpilation bug, because I get an AST validation error with this.

https://closure-compiler-debugger.appspot.com/#input0%3Dfunction%2520random_function()%2520%257B%250A%2520%2520%2520%2520var%2520content%2520%253D%2520%257Blength%253A%25203%257D%253B%250A%2520%2520%2520%2520for%2520(let%2520k%2520%253D%25200%252C%2520%257B%2520length%2520%257D%2520%253D%2520content%253B%2520k%2520%253C%2520length%253B%2520k%252B%252B)%2520%257B%2520%2520%250A%2520%2520%2520%2520%257D%250A%257D%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3Dtrue%26CHECK_SYMBOLS%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CHECK_TYPES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue

I suspect that destructuring transpilation is breaking the `let` statement in the for-loop initialization into multiple `var` statements, which is invalid.

**Comment by brad4d** [comment]
The unit tests for destructuring don't have any for-loop cases with `let` or with multiple declared variables.

https://github.com/google/closure-compiler/blob/master/test/com/google/javascript/jscomp/Es6RewriteDestructuringTest.java

**Comment by lauraharker** [comment]
FYI the above code now throws an error saying we can't transpile it instead of just crashing.

I'm not aware of plans to support transpiling this pattern, though.

https://closure-compiler-debugger.appspot.com/#input0%3Dfunction%2520random_function()%2520%257B%250A%2520%2520%2520%2520var%2520content%2520%253D%2520%257Blength%253A%25203%257D%253B%250A%2520%2520%2520%2520for%2520(let%2520k%2520%253D%25200%252C%2520%257B%2520length%2520%257D%2520%253D%2520content%253B%2520k%2520%253C%2520length%253B%2520k%252B%252B)%2520%257B%2520%2520%250A%2520%2520%2520%2520%257D%250A%257D%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3Dtrue%26CHECK_SYMBOLS%3Dtrue%26MISSING_PROPERTIES%3Dtrue%26TRANSPILE%3Dtrue%26CHECK_TYPES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue

**Comment by concavelenz** [comment]
Perhaps file a new bug for this so that the problem is clear from the description.

**Comment by lauraharker** [comment]
Opened https://github.com/google/closure-compiler/issues/3088 to track implementing transpilation of this pattern.
